crates/providers/src/heart.rs

```diff
@@ -0,0 +1,276 @@
+#![allow(dead_code, unreachable_pub)] // TODO: remove
+//! Block Hearbeat and Transaction Watcher
+
+use alloy_primitives::{B256, U256};
+use alloy_rpc_types::Block;
+use alloy_transport::{utils::Spawnable, TransportErrorKind, TransportResult};
+use futures::{stream::StreamExt, FutureExt, Stream};
+use std::{
+    collections::{BTreeMap, HashMap},
+    fmt,
+    future::Future,
+    time::{Duration, Instant},
+};
+use tokio::{
+    select,
+    sync::{mpsc, oneshot, watch},
+};
+
+/// A configuration object for watching for transaction confirmation.
+pub struct WatchConfig {
+    /// The transaction hash to watch for.
+    tx_hash: B256,
+
+    /// Require a number of confirmations.
+    confirmations: u64,
+
+    /// Optional timeout for the transaction.
+    timeout: Option<Duration>,
+}
+
+impl WatchConfig {
+    /// Create a new watch for a transaction.
+    pub fn new(tx_hash: B256) -> Self {
```

Comment:
we can do `impl From<B256>` so the other functions can accept `impl Into<WatchConfig>`